### PR TITLE
Fix for getCraftRecipes implementation is very slow

### LINKED ISSUE #2210: getCraftRecipes implementation is very slow
The function [getCraftRecipes](https://github.com/minetest/minetest/blob/70074800a20/src/craftdef.cpp#L974) is very slow, and could do better than O(n) by eg. using hashtables. I've created a [minimal test](https://gist.github.com/est31/f4ad8d4403f29ca230e3) using conditions encountered by the dreambuilder game and the unified inventory. This issue causes unified inventory to load ~18 secs at the start of a game. During this time, no client can connect.

### PR DESCRIPTION
**Pull request by gregorycu**
Feedback in GitHub only please.
This is to fix #2210
Fix for minetest.get_craft_recipe
Speed up minetest.get_craft_recipe and minetest.get_all_craft_recipes by several orders of magnitude

### CODE REVIEW

File: src/craftdef.cpp [hunk 1]
```diff
@@ -975,37 +975,20 @@ class CCraftDefManager: public IWritableCraftDefManager
...
-						<<def->dump()<<std::endl;
-				// then go on with the next craft definition
-			}
+		for(std::vector<CraftDefinition*>::const_reverse_iterator
+			i = vecIter->second.rbegin();
+			i != vecIter->second.rend(); i++)
+		{
+			recipes_list.push_back(*i);
 		}
+
```
Comment (ShadowNinja) [inline src/craftdef.cpp]: `get_all_craft_recipes` doesn't specify an order for the results, and `get_craft_recipe` can easily be changed to work with the opposite order if needed.  If the order is definitely important it might be better to store the recipes in reverse order so that in either case you can use a simple copy, or better yet return a const reference.  You should at least `.reserve()` the right amount of space if that won't work though.

Comment (est31) [inline src/craftdef.cpp]: I'm not 100% sure what you mean by this comment, but the order is relevant. For keeping behaviour the same across versions, it needs to be the same as before the change. This is because later registers can override previous ones, and get_all_craft_recipes just dumps all recipes without checking whether a craft occurs twice, just with a changed output number. Usual use case is disabling crafts done by upstream mods.

Comment (gregorycu) [inline src/craftdef.cpp]: So, the reason for this patch isn't to change behaviour or make it more flexible, but if that adds a lot of value, I'll do that.
I think going forward we should be populating lua tables with this data, and mod users should be using that (and having their own caches if they want).
So, I'm not too sure if there is plenty of benefit in adding functionality to this function.

Comment (est31) [inline src/craftdef.cpp]: Tables would break stuff. I'm ok with adding them, but I'd be already satisfied when there would be an additional function `get_actual_craft_recipes` which removes craft duplicate recipes from its output, using the last specified one, like `get_craft_recipe` [does](https://github.com/gregorycu/minetest/blob/Issue-2210/src/craftdef.cpp#L902-903). On the long term however, I think that tables are best.

Comment (gregorycu) [inline src/craftdef.cpp]: Sorry, I meant keep the existing function, but introduce the tables. Yeah, I don't want to break stuff. If we have a slow function, I want to fix that and make it good. If you don't mind, can we keep this issue about the slow perf, and I'll do any enhancements in a different issue?

Comment (est31) [inline src/craftdef.cpp]: OK, just one thing: is it also possible to speed up `get_craft_recipe` too?

Comment (gregorycu) [inline src/craftdef.cpp]: I'll look at it in 8 hours.

File: src/craftdef.cpp [hunk 2]
```diff
@@ -1020,11 +962,17 @@ class CCraftDefManager: public IWritableCraftDefManager
...
 	{
 		verbosestream<<"registerCraft: registering craft definition: "
 				<<def->dump()<<std::endl;
 		m_craft_definitions.push_back(def);
+
+		CraftInput input;
+		ItemStack output;
+		output.deSerialize(def->getOutput(input, gamedef).item);
+
+		m_output_to_craft_definition_map[output.name].push_back(def);
```
Comment (Zeno-) [inline src/craftdef.cpp]: It would seem that these should be cleaned up in ::clear() in the same way m_craft_definitions is.

Comment (gregorycu) [inline src/craftdef.cpp]: Thank you, I changed to code to reflect this comment.

**Comment by est31** [comment]
See this first test run by me: http://pastebin.com/W9Jxbcib
TLDR: There is a new CRASH.

**Comment by gregorycu** [comment]
Hooray for crashing, I'll check this out later tonight.

**Comment by est31** [comment]
Another problem: http://pastebin.com/hyXLvKDc

**Comment by t4im** [comment]
Overwriting crafting recipes via redefining the same crafting input is currently the only way to deactivate/defuse or redefine crafts. Technic is using this as well in a few places: https://github.com/minetest-technic/technic/commit/7e92fe0a46bdc97c39129918ddc226d03862f68d
(and did before that commit similarly, but with actually getting an inventory DoS item)
So unless theres an alternative way added, this behavior should probably be kept ;-)

**Comment by gregorycu** [comment]
Thank you t4im, guess I better make this work then ;)

**Comment by gregorycu** [comment]
I have fixed up the code. It now returns what it did before, with the exception that get_craft_recipe now also has the output.

**Comment by est31** [comment]
Unforfunately the second test with `get_all_craft_recipes` still doesn't have the desired output (should be same as the one seen in "OLD"). But yes I can confirm that `get_craft_recipe` works now, as when I replace the `'funnystone:funnystone 0'` with `'funnystone:funnystone 3'` that craft is chosen.

**Comment by gregorycu** [comment]
I don't have my test output with me, what is missing?

**Comment by est31** [comment]
Nothing changed from http://pastebin.com/hyXLvKDc.

**Comment by gregorycu** [comment]
Thanks est31. I remember that NEW looked the same as OLD apart from the ordering of the fields... I'll doube-check tonight.

**Comment by est31** [comment]
Oh sorry, `get_all_craft_recipes` works fine, just `get_craft_recipe` outputs the craft with 'funnystone:funnystone 0' --  thats wrong.

**Comment by gregorycu** [comment]
Isn't that what happens now?

**Comment by est31** [comment]
Now you get the "group:wood" craft.

**Comment by est31** [comment]
Damn, you were right, now this happens too. PR passes all my tests now :).

**Comment by gregorycu** [comment]
Thanks est31, I want to add proper deregistration functions, and think about how we can improve the existing functions, but that will be a new PR.

Current PR is good to go.

File: src/craftdef.cpp (hunk 1 above)
Comment (ShadowNinja) [inline src/craftdef.cpp]: You should still `.reserve()` space in the new vector, and possibly insert into the front at registration are return a const reference to the vector in the map.

**Comment by est31** [comment]
Out of technical reasons, this has to be rebased, but there is no functional requirement. The lines that got fixed by d902bd31c4b9e8c2a0d85af831a5a23f827a48a2 get removed by this PR anyway, and this also fixes the issue #2222 that the commit fixed.

**Comment by kwolekr** [comment]
Is this PR actually needed as of d902bd3?  It seems like an awful lot of refactoring for a fix that could have been achieved in 2 lines (like d902bd3 does) and the code changed in l_craft.cpp looks a bit wonky (i.e. if it was intended to deduplicate code, why not write a "push_craft_recipe" function instead of what's currently there with the multipleReturn parameter and all that?)

**Comment by est31** [comment]
The advantage is the speed. See #2210 , server instances with the unified inventory need 20 seconds to load, and thats not because its implemented badly, I opened that issue after trying to improve the mod. The method is just called for every registered node, as there is no way to get a table of all craft definitions.

**Comment by kwolekr** [comment]
Oh... I was working under the assumption this fixed something different.  This can wait until after release...

**Comment by t4im** [comment]
d902bd3 fixed #2222; this pr was originally geared towards #2210 then probably thought to fix #2222 as well when working on the same file anyway.

**Comment by gregorycu** [comment]
The PR was for #2210 however I realised I had inadvertently fixed #2222 in the process.

**Comment by gregorycu** [comment]
I have rebased this.

**Comment by kwolekr** [comment]
Since this doesn't fix the API anymore, I think the commit message should be something more telling of what it actually does.
I also wish the code structure could be cleaned up, especially get_craft_recipe_helper.  Why does this single function do the work of both APIs and take that boolean parameter?  The API-specific code should go into each API-specific handler while the common part between both, in this case it seems to be pushing a CraftOutput onto the lua stack, should be the thing separated into its own function.  And what's up with the anonymous namespace?  I get that it has to do with restricting linkage to that translation unit, however I can't help but wonder if it actually has any utility at all here.

This last nag isn't too important, but variable names seem to be camelCase.  I don't think this goes against the code style guidelines but it doesn't quite fit in with the rest.

**Comment by gregorycu** [comment]
With regards to the camcelCase, I can fix that to make it like the rest of the code.
Anonymous namespace is just habit, it prevents clashes (as you know). I'm the kind of guy the restricts visibility by default. Maybe we can get by without it here.
The APIs are incredibly similar, I could probably rejiggle the code to do what you want, but I don't understand how that would give clearer code or less code duplication.
I will change the PR comment to mention #2210 only.

**Comment by gregorycu** [comment]
Thanks @kwolekr  
I have amended the things you asked for, apart from the refactoring to cleaning up the code structure. If you able able to jot down some pseudo-code, I think that would help me understand what you mean. To me, the code looks quite neat, but I am willing to see how you think it may be improved.

**Comment by Zeno-** [comment]
Changed to low priority until:
> apart from the refactoring to cleaning up the code structure
is addressed

**Comment by nerzhul** [comment]
@gregorycu please fix the coding style, like in the previous PR i mention it

**Comment by gregorycu** [comment]
What is wrong with the coding style? Can someone please explain to me how they thing this should look?

**Comment by gregorycu** [comment]
I won't be able to fix the code structure until someone clearly explains what is wrong with it.

**Comment by gregorycu** [comment]
I have moved the brackets onto the previous line. If someone knows what is wrong with the code structure, can they please indicate what that is?

**Comment by nerzhul** [comment]
Code seems good for me

File: src/craftdef.cpp [hunk 3]
```diff
@@ -930,84 +930,24 @@ class CCraftDefManager: public IWritableCraftDefManager
...
-				errorstream<<"getCraftResult: ERROR: "
-						<<"Serialization error in recipe "
-						<<def->dump()<<std::endl;
-				// then go on with the next craft definition
-			}
-		}
-		return false;
-	}
 	virtual std::vector<CraftDefinition*> getCraftRecipes(CraftOutput &output,
 			IGameDef *gamedef) const
```
Comment (ShadowNinja) [inline src/craftdef.cpp]: As an optimization, this should accept a `only_first` or `limit` argument.

File: src/craftdef.cpp [hunk 4]
```diff
@@ -1022,11 +962,19 @@ class CCraftDefManager: public IWritableCraftDefManager
...
 				<<def->dump()<<std::endl;
 		m_craft_definitions.push_back(def);
+
+		CraftInput input;
+		std::string outputItem;
+
+		std::istringstream ss (def->getOutput(input, gamedef).item);
+		ss >> outputItem;
+
+		m_output_to_craft_definition_map[outputItem].push_back(def);
```
Comment (ShadowNinja) [inline src/craftdef.cpp]: This won't work with old-style quoted item strings (eg `"default:stone" 5`).

File: src/script/lua_api/l_craft.cpp [hunk 5]
```diff
@@ -326,118 +326,87 @@ int ModApiCraft::l_get_craft_result(lua_State *L)
...
+		lua_setfield(L, -2, "output");
+
+		if (return_multiple)
+			lua_rawseti(L, -2, ++i);
+		else
+			break;
 	}
+
 	return 1;
 }
```
Comment (ShadowNinja) [inline src/script/lua_api/l_craft.cpp]: I think a cleaner way to design this is with `push_craft_recipe` and `push_craft_recipes` functions, the latter would call the former in a loop and the former could be used directly.

File: src/craftdef.cpp [hunk 6]
```diff
@@ -930,84 +930,24 @@ class CCraftDefManager: public IWritableCraftDefManager
...
-				}
-			}
-			catch(SerializationError &e)
-			{
-				errorstream<<"getCraftResult: ERROR: "
-						<<"Serialization error in recipe "
-						<<def->dump()<<std::endl;
-				// then go on with the next craft definition
-			}
+		for(std::vector<CraftDefinition*>::const_reverse_iterator
```
Comment (ShadowNinja) [inline src/craftdef.cpp]: You should `recipes_list.reserve(vec_iter->second.size());`

**Comment by gregorycu** [comment]
Not likely to ever be merged. Closing.

**Comment by ShadowNinja** [comment]
Not a valid closing reason.

**Comment by est31** [comment]
I think he should be abled to close the PR when he doesn't want to contribute on it anymore. His work is well documented and published, in a way another contributor can easily complete it. But that would require another PR out of the technical reason that a PR's branches can't be changed.

**Comment by ShadowNinja** [comment]
@est31 This does need some work, but if it's closed it will be forgotten, not fixed up.

**Comment by gregorycu** [comment]
I'm looking at the 98 outstanding pull requests. I don't think keeping it open makes a difference. If core devs spent less time insulting me, and more time merging, the project may be in a better place. And maybe I'd be inclined to continue contributing.

**Comment by est31** [comment]
One of your pull requests were merged yesterday. Mine get merged quite fast. I'm sure when you are nice in the comments and fix the problems, your PR will get merged.

**Comment by proller** [comment]
got it in 2 months. faster and faster..

**Comment by hunter688** [comment]
It is tragic that a pull request with the code to improve the quality of the game have a low priority.

**Comment by Zeno-** [comment]
Maybe we should get rid of the low priority label. It seems kind of useless IMO anyway

**Comment by kwolekr** [comment]
https://github.com/minetest/minetest/commit/03e0dd33a847a83d975282c6caf6b926306e7b57
